### ISSUE TITLE
Random forest with weighted sampling

### ISSUE THREAD

**Issue by albertmolinermrf** [haifengl/smile#280]
When training sets have repetitive items and the overall amount of data is huge, it would be useful to pass an optional, extra array of weights to RandomForest constructor telling the (relative) frequency of each item. The result should be equivalent to passing a training set with the items repeated. For example, the two calls would be equivalent:
```java
  double[][] x = new double[][] {
    {1.0, 1.0, 1.0},
    {1.0, 1.0, 1.0},
    {1.0, 1.0, 1.0},
    {2.0, 2.0, 2.0},
  };
  double[] y = new double[] {2.0, 2.0, 2.0, 4.0};
  RandomForest randomForest = new RandomForest(x, y, ...);
```
```java
  double[][] x = new double[][] {
    {1.0, 1.0, 1.0},
    {2.0, 2.0, 2.0},
  };
  double[] y = new double[] {2.0, 4.0};
  double[] weights = new double[] {3.0, 1.0};  // {0.3, 0.1} would have the same effect
  RandomForest randomForest = new RandomForest(x, y, ..., weights);
```
If the weights are applied at sampling level, the advantage in memory and training time can be very significant when the training set is very big and some frequencies are much higher than others.
It may be argued that the `subsample` parameter can be tweaked to reduce the amount of items sent to each tree and therefore to reduce the resources consumed, but there is still a physical limitation: Java arrays cannot grow over `Integer.MAX_VALUE`. Some training sets are really that big but fortunately can be aggregated with equal rows, with some aggregations having more frequency than others.
Probably this functionality, if accepted, could be implemented by slightly modifying the sampling code in `RandomForest.TrainingTask`. Do you want us to provide a pull request with a proposal of code?

**Comment by haifengl** [comment]
Is this only for regression? The repetitive always have the same response values? Thanks.

**Comment by haifengl** [comment]
The weight will have big impact on the sampling strategy and approximate split point determination. What's your algorithm on these points?

**Comment by albertmolinermrf** [comment]
> Is this only for regression?

No, it could be applied to classification too.

> The repetitive always have the same response values?

Yes.

> The weight will have big impact on the sampling strategy and approximate split point determination. What's your algorithm on these points?

It would be applied only by modifying the sampling probability of each data point. This can be easily achieved by:

1. Creating an array of accumulated weights (if no weights array is passed, all weights are taken to be 1).
1. Generating a random number between 0 and the last accumulated value.
1. Picking the data point at the position where the random number above would fall if it had to be inserted into the accumulation array.

This way, data points with higher weight would have higher probability during sampling.

Split point determination would not be affected by this approach. Therefore, loss function of each tree would not change, but loss function of the whole forest would be biased by the weights due to the bias introduced in the sampling.

**Comment by haifengl** [comment]
Weighted sampling is actually problematic. Suppose the weight/frequency of a sample is 10, should we treat it as 10 samples selected if it is sampled? It is definitely very different behavior from the standard sampling. The policy to determine the frequency of selected samples should depends on the data. It is hard for us to implement the proper policy inside the algorithm.

Besides, if the data have many duplicated samples, some (down-) sampling should be performed first. I suggest the users to do it.
